File: app/src/main/java/com/bigscreen/mangindo/network/loader/NewReleaseLoader.java
package com.bigscreen.mangindo.network.loader;

import com.bigscreen.mangindo.network.NetworkError;
import com.bigscreen.mangindo.network.model.response.NewReleaseResponse;
import com.bigscreen.mangindo.network.service.MangaApiService;

import rx.Subscription;
import rx.subscriptions.CompositeSubscription;

public class NewReleaseLoader {

    private MangaApiService mangaApiService;
    private OnLoadNewReleaseListener loadNewReleaseListener;
    private CompositeSubscription subscriptions;

    public NewReleaseLoader(MangaApiService apiService, OnLoadNewReleaseListener listener) {
        mangaApiService = apiService;
        loadNewReleaseListener = listener;
        subscriptions = new CompositeSubscription();
    }

    public void loadNewReleaseManga() {
        loadNewReleaseListener.onPrepareLoadData();
        Subscription subscription = mangaApiService.getNewReleaseManga(new MangaApiService.LoadMangaListCallback() {
            @Override
            public void onSuccess(NewReleaseResponse response) {
                if (response != null)
                    loadNewReleaseListener.onSuccessLoadData(response);
                else
                    loadNewReleaseListener.onFailedLoadData("Manga tidak dapat ditemukan");
            }

            @Override
            public void onError(NetworkError networkError) {
                loadNewReleaseListener.onFailedLoadData(networkError.getErrorMessage());
            }
        });
        subscriptions.add(subscription);
    }

    public void unSubscribe() {
        subscriptions.unsubscribe();
    }

    public interface OnLoadNewReleaseListener {
        void onPrepareLoadData();

        void onSuccessLoadData(NewReleaseResponse newReleaseResponse);

        void onFailedLoadData(String message);
    }
}

File: app/src/main/java/com/bigscreen/mangindo/newrelease/NewReleaseLoader.kt
package com.bigscreen.mangindo.newrelease

import com.bigscreen.mangindo.common.extension.subscribes
import com.bigscreen.mangindo.network.model.response.NewReleaseResponse
import com.bigscreen.mangindo.repos.MangaRepository

import rx.subscriptions.CompositeSubscription

class NewReleaseLoader(
        private val repository: MangaRepository,
        private val listener: OnLoadNewReleaseListener
) {

    private val subscriptions = CompositeSubscription()

    fun loadNewReleaseManga() {
        listener.onPrepareLoadData()
        val subscription = repository.getNewReleased().subscribes(
                { listener.onSuccessLoadData(it) },
                { listener.onFailedLoadData(it.getErrorMessage()) }
        )
        subscriptions.add(subscription)
    }

    fun unSubscribe() {
        subscriptions.unsubscribe()
    }

    interface OnLoadNewReleaseListener {

        fun onPrepareLoadData()

        fun onSuccessLoadData(newReleaseResponse: NewReleaseResponse)

        fun onFailedLoadData(message: String)
    }
}
